File: algorithm/leetcode1014_best-sightseeing-pair.cpp
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& A) {
        int i, j, scores = 0;
        for(i = 0; i < A.size(); ++i)
        {
        	for(j = i+1; j < A.size(); ++j)
        		scores = max(scores, A[i]+A[j]+i-j);
        }
        return scores;
    }
};
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& A) {
        int i, scoresA = A[0]+0, scores = 0;
        for(i = 1; i < A.size(); ++i)
        {
        	scores = max(scores, scoresA + A[i]-i);//历史的A的得分，当前的B得分
        	scoresA = max(scoresA, A[i]+i); //记录最大的A点得分
        }
        return scores;
    }
};

File: algorithm/Python3_version/leetcode1014.best-sightseeing-pair.py
class Solution:
    def maxScoreSightseeingPair(self, A: List[int]) -> int:
        scoreA = A[0]+0
        score = 0
        for i in range(1, len(A)):
            score = max(score, scoreA+A[i]-i)
            scoreA = max(scoreA, A[i]+i)
        return score